src/linker/Linker.Steps/DiscoverSerializationHandler.cs

```diff
@@ -0,0 +1,144 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Mono.Cecil;
+using Mono.Linker.Dataflow;
+using Mono.Linker.Steps;
+
+namespace Mono.Linker.Steps
+{
+	// This discovers types attributed with certain serialization attributes, to match the old behavior
+	// of xamarin-android. It is not meant to be complete. Unlike xamarin-andorid:
+	// - this will only discover attributed types that are marked
+	// - this will discover types in non-"link" assemblies as well
+	public class DiscoverSerializationHandler : IMarkHandler
+	{
+		LinkContext _context;
+
+		public void Initialize (LinkContext context, MarkContext markContext)
+		{
+			_context = context;
+			markContext.RegisterMarkTypeAction (type => ProcessType (type));
+			markContext.RegisterMarkMethodAction (method => CheckForSerializerActivation (method));
+		}
+
+		void CheckForSerializerActivation (MethodDefinition method)
+		{
+			var type = method.DeclaringType;
+
+			if (!_context.SerializationMarker.IsActive (SerializerKind.DataContractSerializer) &&
+				method.IsConstructor && !method.IsStatic &&
+				type.Namespace == "System.Runtime.Serialization" &&
+				type.Name == "DataContractSerializer") {
```

Comment:
What about `DataContractJsonSerializer` (note that it doesn't derive from `DataContractSerializer`)?